client/tracing/src/block/mod.rs

```diff
@@ -0,0 +1,253 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utilities for tracing block execution
+
+use std::collections::HashMap;
+use std::sync::{Arc};
+use std::sync::atomic::{AtomicU64, Ordering};
+
+use parking_lot::Mutex;
+use tracing::{Dispatch, dispatcher, Subscriber};
+use tracing_subscriber::CurrentSpan;
+
+use sc_client_api::BlockBackend;
+use sp_api::{Core, Metadata, ProvideRuntimeApi};
+use sp_blockchain::HeaderBackend;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, Header},
+};
+use sp_rpc::tracing::{BlockTrace, Span, Event, Values};
+use sp_tracing::{WASM_NAME_KEY, WASM_TARGET_KEY, WASM_TRACE_IDENTIFIER};
+
+use tracing_core::span::{Attributes, Record, Id};
+use tracing_core::{Level};
+use std::time::{Instant};
+
+// Default to only runtime and state related traces
+const DEFAULT_TARGETS: &'static str = "pallet,frame,sp_io::storage=debug";
+const TRACE_TARGET: &'static str = "block_trace";
+
+struct BlockSubscriber {
+	targets: Vec<(String, Level)>,
+	next_id: AtomicU64,
+	current_span: CurrentSpan,
+	spans: Arc<Mutex<HashMap<Id, Span>>>,
+	events: Arc<Mutex<Vec<Event>>>,
```

Comment:
Why are these `Arc`s?

Reply:
`tracing::Dispatch` takes an owned `Subscriber` with constraints `S: Subscriber + Send + Sync + 'static, ` so `Rc` doesn't work.

The [way we capture the traces from a single block](https://github.com/paritytech/substrate/pull/7780/files#diff-f5ee9da45314c57567317ca337c1686d02bcf4d13779775ec99186d275e51957R196-R206) is through a new local Dispatcher within which we execute the block; then we drop the `Dispatcher` and unwrap the `Arc` to get the spans and events.

I don't see a better way, do you?

Reply:
Ahh I had overseen this.

There would another way by using this: https://docs.rs/tracing/0.1.22/tracing/struct.Dispatch.html#method.downcast_ref

Reply:
That's a good idea; I tried this [here](https://github.com/paritytech/substrate/pull/7780/commits/d9fd09d33cfcffe005c9c82d37a0668e04844891) but there's [a test](https://github.com/paritytech/substrate/blob/5c183f3b1110e1cefdc74a2f8787187d6080ffa8/client/tracing/src/lib.rs#L630) that is behaving weirdly with this code. Not sure if it's the code or the test (or both).
`cargo t -p sc-tracing` *skips* the test for some reason, but filtering for it reveals the failure (`cargo t -p sc-tracing thread`).
TBH I'm not sure if the test is correct or not.

Reply:
This test is failing or what? :D

Reply:
It is failing. But as you see all tests pass in CI here, so wtf? Turns out that the tests in `logging/`, e.g. [here](https://github.com/paritytech/substrate/blob/5c183f3b1110e1cefdc74a2f8787187d6080ffa8/client/tracing/src/logging/mod.rs#L389-L394) that execute out-of-process somehow swallow the other tests; I'm guessing it messes up `stderr`. 
`cargo t -p sc-tracing` claims to run 14 tests, but actually just runs 8. Sometimes 9 (the one failing), and when it does, the test fails but does not report the failure to the test runner, so it's like it was never there.

So I have two problems:
1. Fix the out-of-process tests (I'll try moving them to a crate and see if that helps; cc @cecton)
2. Figure out if the failing tracing test should fail or pass. A question I have here is if, in your opinion, that test is correct or not.